miden-lib/asm/note_scripts/basic.masm

```diff
@@ -0,0 +1,126 @@
+#! 
+#! Requires that the account exposes: miden::wallets::basic::receive_asset procedure.
+#!
+#! Inputs: [target_account_id]
+#! Outputs: []
+#!
+#! - ASSET is the asset to be received, can be fungible or non-fungible
+#!
+#! FAILS if: 
+#! - Account does not expose miden::wallets::basic::receive_asset procedure
+#! - Account ID of executing account is not equal to specified Account ID
+#! - The same non-fungible asset already exists in the account.
+#! - Adding a fungible asset would result in amount overflow, i.e.,
+#!   the total amount would be greater than 2^63.
+export.p2id                                             
+    # => [note_inputs = target_account_id, ...]                                            
+    
+    exec.account::get_id                                
+    # => [account_id, target_account_id, ...]
+    
+    # ensure account_id = target_account_id, fails otherwise
+    assert_eq                                           
+    # => [...]
+    
+    exec.receive_assets
+    # => [...]
+end
+
+#! Pay to ID reclaimable - ensures that note can only be consumed by an account with the provided ID
+#! or after provided block number also by the sender account
+#! 
+#! Inputs: # [reclaim_block_height, target_account_id, ...]
+#! Outputs: []
+#!
+#! - ASSET is the asset to be received, can be fungible or non-fungible
+#!
+#! FAILS if: 
+#! - Account does not expose miden::wallets::basic::receive_asset procedure
+#! - Account ID of executing account is not equal to specified Account ID
+#! - The same non-fungible asset already exists in the account.
+#! - Adding a fungible asset would result in amount overflow, i.e.,
+#!   the total amount would be greater than 2^63.
+export.p2idr
+    # => [reclaim_block_height, target_account_id, ...]                                                                                        
+    
+    exec.tx::get_block_number                           
+    # => [current_block_height, reclaim_block_height, target_account_id, ...]    
+    
+    lte                                                 
+    # => [1 || 0, target_account_id, ...], 1 if current_block_height >= reclaim_block_height, 0 otherwise
+
+    # current_block_height >= reclaim_block_height
+    if.true                                             
+    # => [target_account_id, ...]
+        
+        exec.account::get_id dup
+        # => [account_id, account_id, target_account_id, ...]
+        
+        exec.note::get_sender
+        # => [sender_account_id, account_id, account_id, target_account_id, ...]
+        
+        # fails if account_id is neither sender nor target
+        eq swap.2 eq add assert
+        
+    # current_block_height < reclaim_block_height
+    else
+    # => [target_account_id, ...]
+        exec.account::get_id                            
+        # => [account_id, target_account_id, ...]
+
+        assert_eq
+        # => fails if account_id != target_account_id
+    end
```

Comment:
I would write this script something like this:
```
exec.account::get_id dup
# => [account_id, account_id, reclaim_block_height, target_account_id, ...]

# determine if the current account is the target account
movup.3 eq
# => [is_target, account_id, reclaim_block_height, ...]

if.true
    # if current account is the target, we don't need to check anything else,
    # and so we just clear the stack
    drop drop
else
    # if the current account is not the target, we need to make sure that it is
    # the sender
    exec.note::get_sender assert_eq
    # => [reclaim_block_height, ...]

    # and that the current block height is greater than or equal to the the
    # reclaim_block_height
    exec.tx::get_block_number u32checked_lte assert
end
```

The reasons are:
- `lte` is relatively more expensive than `u32checked_lte` (18 vs. 8 cycles), and since block height should never be greater than `u32`, we should use `u32checked_lte`.
- Since even `u32checked_lte` takes 8 cycles, we should try to avoid doing it in the most likely case (e.g., when the recipient claims the note). So, I'd check for the recipient first, and only if the sender executes the transaction, check that the block height valid.